R/valid.R

```diff
@@ -1,41 +1,114 @@
-#' Validation of R7 objects
+#' Validate an R7 object
 #'
-#' [validate()] calls the validation of an R7 object. This is done
-#' automatically when creating new objects (at the end of [new_object]) and
-#' when setting any property.
+#' @description
+#' `validate()` calls the validator of an R7 object. This is done automatically
+#' when creating new objects (at the end of [new_object()]) and when setting
+#' any property with [prop<-].
 #'
-#' [valid_eventually()] disables validation of properties, runs a function on
-#' the object, then validates the object.
+#' `valid_eventually()` disables validation, modifies the object, then
+#' revalidates. This is useful when a sequence of operations would otherwise
+#' lead an object to be temporarily invalid.
 #'
-#' [valid_implicitly()] does the same but does not validate the object at the end.
-#'
-#' [valid_implicitly()] should only be used rarely in performance critical code
-#' where you are certain a sequence of operations cannot produce an invalid
-#' object.
+#' `valid_implicitly()` does the same but does not validate the object at the
+#' end. It should only be used rarely, and in performance critical code where
+#' you are certain a sequence of operations cannot produce an invalid object.
 #' @param object An R7 object
 #' @param fun A function to call on the object before validation.
-validate <- function(object) {
+#' @param properties If `TRUE`, the default, checks property types before
+#'   executing the validator.
+#' @export
+#' @examples
+#' # A range class might validate that the start is less than the end
+#' Range <- new_class("Range",
+#'   properties = list(start = "double", end = "double"),
+#'   validator = function(object) {
+#'     if (object@start >= object@end) "start must be smaller than end"
+#'   }
+#' )
+#' # You can't construct an invalid object:
+#' try(Range(1, 1))
+#'
+#' # And you can't create an invalid object with @<-
+#' r <- Range(1, 2)
+#' try(r@end <- 1)
+#'
+#' # But what if you want to move a range to the right?
+#' rightwards <- function(r, x) {
+#'   r@start <- r@start + x
+#'   r@end <- r@end + x
+#'   r
+#' }
+#' # This function doesn't work because it creates a temporarily invalid state
+#' try(rightwards(r, 10))
+#'
+#' # This is the perfect use case for valid_eventually():
+#' rightwards <- function(r, x) {
+#'   valid_eventually(r, function(object) {
+#'     object@start <- object@start + x
+#'     object@end <- object@end + x
+#'     object
+#'   })
+#' }
+#' rightwards(r, 10)
+#'
+#' # Alternatively, you can set multiple properties at once using props<-
+#' rightwards <- function(r, x) {
+#'   props(r) <- list(start = r@start + x, end = r@end + x)
+#'   r
+#' }
+#' rightwards(r, 20)
+validate <- function(object, properties = TRUE) {
```

Comment:
Yeah, agreed with both points. But since this is going in base R (which doesn't use that pattern), and `properties` is mostly their for internal use, I'll leave as is.

Reply:
Adding an unneeded `...` to the argument list also comes with a price.  I would never want to pay that just to force users..

Reply:
Explicit formal arguments also improve code clarity. Users are free to follow whatever calling convention they like.